broker.go

```diff
@@ -0,0 +1,584 @@
+// and creates a stateful message handler for each message source peer.  An
+// optional external policy determines whether to allow or deny handling
+// messages from specific peers.
+//
+// Messages to separate peers are handled concurrently, and multiple messages
+// to the same peer are handled serially.  If a handler is busy handling a
+// message, and more messages arrive for the same peer, then the last message
+// replaces the previous unhandled message to avoid having to maintain queues
+// of messages.  Handlers do not have persistent goroutines, but start a new
+// goroutine to handle a single message.
+type LegBroker struct {
+	dtManager   dt.Manager
+	gsExchange  graphsync.GraphExchange
+	tmpDir      string
+	unsubEvents dt.Unsubscribe
+
+	host    host.Host
+	psub    *pubsub.Subscription
+	topic   *pubsub.Topic
+	addrTTL time.Duration
+
+	// dss captures the default selector sequence passed to ExploreRecursiveWithStopNode
+	dss ipld.Node
+
+	handlers      map[peer.ID]*handler
+	handlersMutex sync.Mutex
+	peerPolicy    PeerPolicyHandler
+
+	// Map of CID of in-progress sync to sync done channel.
+	syncDoneChans map[cid.Cid]chan<- struct{}
+	syncDoneMutex sync.Mutex
+
+	// inEvents is used to send a ChangeEvent from a peer handler to the
+	// distributeEvents goroutine.
+	inEvents chan ChangeEvent
+
+	// outEventsChans is a slice of channels, where each channel delivers a
+	// copy of a ChangeEvent to an OnChange reader.
+	outEventsChans []chan ChangeEvent
+	outEventsMutex sync.Mutex
+
+	// cancelBroker cancels broker goroutines
+	cancelBrkr context.CancelFunc
+	// cancelAll cancels pubsub and datatransfer and broker goroutines
+	cancelAll context.CancelFunc
+}
+
+// ChangeEvent notifies an OnChange reader that a specified peer completed a sync.
+type ChangeEvent struct {
+	Cid    cid.Cid
+	PeerID peer.ID
+}
+
+// Defined tells whether the ChangeEvent holds any data.
+func (ce ChangeEvent) Defined() bool {
+	return ce.Cid.Defined()
+}
+
+// handler holds state that is specific to a peer
+type handler struct {
+	dtManager dt.Manager
+	// distEvents is used to communicate a syncDone event back to the
+	// LegBroker for distribution to OnChange readers.
+	distEvents chan<- ChangeEvent
+	latestSync ipld.Link
+	msgChan    chan cid.Cid
+	// peerID is the ID of the peer this handler is responsible for.
+	peerID peer.ID
+	// putSyncDone is a function supplied by the LegBroker for the handler
+	// to give LegBroker the channel to send sync done notification to.
+	putSyncDone func(cid.Cid, chan<- struct{})
+	// syncMutex serializes the handling of individual syncs
+	syncMutex sync.Mutex
+}
+
+// NewLegBroker creates a new LegBroker, and will process messages that are allowed by given peerPolicy.
+func NewLegBroker(host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string, dss ipld.Node, addrTTL time.Duration, peerPolicy PeerPolicyHandler) (*LegBroker, error) {
```

Comment:
@masih Sure, I will add options.  I was just trying to keep it closer to LegMultisubscriber, but yeah... options would be good.  Done.

@MarcoPolo With some of the changes, it cannot be a drop-in.  It is close enough to make it tempting to try and do that, but I think that would detract from clarity.  I do not see the need for defining an interface, aside from test mocking, since we are only supplying on implementation.  If there was a consumer of multiple implementations, that consumer can define that interface.